### PR TITLE
User-facing changelog for JupyterLab 4.2

### PR DESCRIPTION
**Pull request by krassowski**
## References

Closes https://github.com/jupyterlab/jupyterlab/issues/16076

Review rendered changes here: https://jupyterlab--16103.org.readthedocs.build/en/16103/getting_started/changelog.html

I only documented things I remembered on top of my head. If I missed something please do leave a comment!

This is not ready because we may have more features, like:
- https://github.com/jupyterlab/jupyterlab/pull/15155
- https://github.com/jupyterlab/jupyterlab/pull/15623
- a section on improved keyboard shortcuts:
   - https://github.com/jupyterlab/jupyterlab/pull/16078
   - https://github.com/jupyterlab/jupyterlab/pull/15927
- not sure if we want to mention that this release enables server-side execution because this is not implemented yet

## Code changes

None

## User-facing changes

None

## Backwards-incompatible changes

None

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -4,6 +4,38 @@
 
 # JupyterLab Changelog
 
+## v4.2
+
+### Workspaces UI
+
+It is now possible to switch and manage [workspaces](https://jupyterlab.readthedocs.io/en/latest/user/workspaces.html) from within JupyterLab.
+
+<img alt="A context menu opened over the sidebar with workspaces list. The menu has options to manager the workspace over which it was opened: clone, rename, reset delete, export and import." src="https://raw.githubusercontent.com/krassowski/jupyterlab/ufc4.2/docs/source/getting_started/changelog_assets/4.2-workspaces-sidebar.png" class="jp-screenshot">
```
Comment (krassowski) [inline CHANGELOG.md:13]: I see this warning on CI but I do not understand what it is trying to tell me:

```
/home/runner/work/jupyterlab/jupyterlab/docs/source/getting_started/changelog.md:13: WARNING: 'image': Invalid options format (bad YAML) [myst.directive_parse]
```

Comment (krassowski) [reply CHANGELOG.md:13]: It does not show up for other `` tags

Comment (krassowski) [reply CHANGELOG.md:13]: Maybe it is because `:` in alt which is making it difficult for `myst`?

**Comment by jtpio** [comment]
> not sure if we want to mention that this release enables server-side execution because this is not implemented yet

Maybe just a small not that server-side execution will be explored in `jupyter-collaboration` after the 4.2 release?

**Comment by krassowski** [comment]
I think this is ready for review, I think there will be no more new features merged but we can confirm this on the lab call tomorrow. We can add bugfixes and enhancements like https://github.com/jupyterlab/jupyterlab/pull/15927 and https://github.com/jupyterlab/jupyterlab/pull/15948 in follow-up PRs.

File: CHANGELOG.md [hunk 2]
```diff
@@ -4,6 +4,55 @@
...
+   <img alt="A modal dialog for quickly jumping to open/closed files" src="https://raw.githubusercontent.com/krassowski/jupyterlab/ufc4.2/docs/source/getting_started/changelog_assets/4.2-recent-modal.png" class="jp-screenshot">
+
+### Full notebook windowing mode by default
+
+Notebooks in the `full` windowing mode only render the visible cells, significantly improving the performance of the application. The limitation of `full` mode is that the search function in your browser may produce false negatives, as discussed [here](https://jupyterlab.readthedocs.io/en/latest/user/interface.html#searching) - using the JupyterLab search function is recommended. To revert to the behaviour from JupyterLab 4.1 go to Settings → Settings Editor → Notebook, scroll to "Windowing mode" and choose `defer`.
+
+### Improved Shortcuts Editor
+
+Among numerous improvements and bug fixes for keyboard shortcuts editor:
+- it is now possible to remove the default shortcuts,
```
Comment (gabalafou) [inline CHANGELOG.md:37]:
```suggestion

- it is now possible to remove the default shortcuts,
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -4,6 +4,55 @@
...
+- shortcuts are correctly sorted when using a language pack,
+- shortcuts with different arguments are now correctly displayed as individual entries.
+
+<img alt="A context menu opened over a keybinding in the Keyboard Shortcuts setting panel within the Settings widget" src="https://raw.githubusercontent.com/krassowski/jupyterlab/ufc4.2/docs/source/getting_started/changelog_assets/4.2-delete-shortcut.png" class="jp-screenshot">
+
+### Dark high contrast theme
+
+A new dark high contrast theme, which is intended to benefit users with the need for higher contrast, following the WCAG AAA accessibility standard for color contrast.
```
Comment (gabalafou) [inline CHANGELOG.md:45]:
```suggestion
A new theme, JupyterLab Dark High Contrast, which is intended to benefit users with the need for higher contrast, following the WCAG AAA accessibility standard for color contrast.
```

File: CHANGELOG.md [hunk 4]
```diff
@@ -4,6 +4,55 @@
...
+To select this theme choose from the menu bar: Settings → Theme → JupyterLab Dark High Contrast. Please do provide feedback and suggestions on further improvements to this theme.
```
Comment (gabalafou) [inline CHANGELOG.md:46]:
```suggestion
To select this theme choose from the menu bar: Settings → Theme → JupyterLab Dark High Contrast. Please provide feedback and suggestions on further improvements to this theme.
```

File: CHANGELOG.md [hunk 5]
```diff
@@ -4,6 +4,55 @@
...
+Users are now able to:
+
+- re-open recently used files from the main menu:
+
+   <img alt="An 'Open Recent' sub-menu in the 'File' menu allowing to re-open recently used files" src="https://raw.githubusercontent.com/krassowski/jupyterlab/ufc4.2/docs/source/getting_started/changelog_assets/4.2-recent-submenu.png" class="jp-screenshot">
+- re-open recently closed files from the sidebar:
+
+   <img alt="A new 'Recently Closed' section in the 'Running' sidebar" src="https://raw.githubusercontent.com/krassowski/jupyterlab/ufc4.2/docs/source/getting_started/changelog_assets/4.2-recent-sidebar.png" class="jp-screenshot">
+- quickly jump to open tabs/recently closed files by using a new searchable modal dialog (press <kbd>Ctrl</kbd>+<kbd>Alt</kbd>+<kbd>A</kbd> to open the dialog, or click on the \[↗\] icon in the sidebar):
+   <img alt="A modal dialog for quickly jumping to open/closed files" src="https://raw.githubusercontent.com/krassowski/jupyterlab/ufc4.2/docs/source/getting_started/changelog_assets/4.2-recent-modal.png" class="jp-screenshot">
```
Comment (gabalafou) [inline CHANGELOG.md:28]: These new lines should make the change log render better in GitHub:

```suggestion
- re-open recently used files from the main menu:

   <img alt="An 'Open Recent' sub-menu in the 'File' menu allowing to re-open recently used files" src="https://raw.githubusercontent.com/krassowski/jupyterlab/ufc4.2/docs/source/getting_started/changelog_assets/4.2-recent-submenu.png" class="jp-screenshot">
   
- re-open recently closed files from the sidebar:

   <img alt="A new 'Recently Closed' section in the 'Running' sidebar" src="https://raw.githubusercontent.com/krassowski/jupyterlab/ufc4.2/docs/source/getting_started/changelog_assets/4.2-recent-sidebar.png" class="jp-screenshot">
   
- quickly jump to open tabs/recently closed files by using a new searchable modal dialog (press <kbd>Ctrl</kbd>+<kbd>Alt</kbd>+<kbd>A</kbd> to open the dialog, or click on the \[↗\] icon in the sidebar):

   <img alt="A modal dialog for quickly jumping to open/closed files" src="https://raw.githubusercontent.com/krassowski/jupyterlab/ufc4.2/docs/source/getting_started/changelog_assets/4.2-recent-modal.png" class="jp-screenshot">
```

File: CHANGELOG.md [hunk 6]
```diff
@@ -4,6 +4,55 @@
...
+
+### Full notebook windowing mode by default
+
+Notebooks in the `full` windowing mode only render the visible cells, significantly improving the performance of the application. The limitation of `full` mode is that the search function in your browser may produce false negatives, as discussed [here](https://jupyterlab.readthedocs.io/en/latest/user/interface.html#searching) - using the JupyterLab search function is recommended. To revert to the behaviour from JupyterLab 4.1 go to Settings → Settings Editor → Notebook, scroll to "Windowing mode" and choose `defer`.
```
Comment (gabalafou) [inline CHANGELOG.md:32]:
```suggestion
Notebooks in the `full` windowing mode only render the visible cells, significantly improving the performance of the application. The limitation of `full` mode is that the search function in your browser may produce false negatives - using the [JupyterLab search function](https://jupyterlab.readthedocs.io/en/latest/user/interface.html#searching) is recommended. To revert to the behaviour from JupyterLab 4.1 go to Settings → Settings Editor → Notebook, scroll to "Windowing mode" and choose `defer`.
```

File: CHANGELOG.md [hunk 7]
```diff
@@ -4,6 +4,55 @@
...
+
+A new dark high contrast theme, which is intended to benefit users with the need for higher contrast, following the WCAG AAA accessibility standard for color contrast.
+To select this theme choose from the menu bar: Settings → Theme → JupyterLab Dark High Contrast. Please do provide feedback and suggestions on further improvements to this theme.
+
+<img alt="A dark high contrast theme" src="https://raw.githubusercontent.com/krassowski/jupyterlab/ufc4.2/docs/source/getting_started/changelog_assets/4.2-dark-high-contrast.png" class="jp-screenshot">
+
+### Keyboard shortcuts improvements
+
+Following feedback from users, implementation of certain shortcuts was moved from the CodeMirror editor to JupyterLab shortcut system, allowing for better integration:
+- the shortcut for deleting a line was restored to <kbd>Ctrl</kbd>+<kbd>D</kbd> as in JupyterLab 3.x; the shortcut for extending (multi-cursor) selection was remapped to <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>D</kbd>. The <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>K</kbd> shortcut can also be used for deleting lines for compatibility with other editors.
```
Comment (gabalafou) [inline CHANGELOG.md:53]:
```suggestion

- the shortcut for deleting a line was restored to <kbd>Ctrl</kbd>+<kbd>D</kbd> as in JupyterLab 3.x; the shortcut for extending (multi-cursor) selection was remapped to <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>D</kbd>. The <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>K</kbd> shortcut can also be used for deleting lines for compatibility with other editors.
```

File: CHANGELOG.md [hunk 8]
```diff
@@ -4,6 +4,55 @@
...
+<img alt="A context menu opened over a keybinding in the Keyboard Shortcuts setting panel within the Settings widget" src="https://raw.githubusercontent.com/krassowski/jupyterlab/ufc4.2/docs/source/getting_started/changelog_assets/4.2-delete-shortcut.png" class="jp-screenshot">
+
+### Dark high contrast theme
+
+A new dark high contrast theme, which is intended to benefit users with the need for higher contrast, following the WCAG AAA accessibility standard for color contrast.
+To select this theme choose from the menu bar: Settings → Theme → JupyterLab Dark High Contrast. Please do provide feedback and suggestions on further improvements to this theme.
+
+<img alt="A dark high contrast theme" src="https://raw.githubusercontent.com/krassowski/jupyterlab/ufc4.2/docs/source/getting_started/changelog_assets/4.2-dark-high-contrast.png" class="jp-screenshot">
+
+### Keyboard shortcuts improvements
```
Comment (gabalafou) [inline CHANGELOG.md:50]: The other headings are more specific so I tried to make this one more specific too

```suggestion
### JupyterLab-ify cell editor (CodeMirror) keyboard shortcuts
```

**Review by gabalafou — COMMENTED** [review]
This is great!

File: CHANGELOG.md (hunk 8 above)
Comment (krassowski) [reply CHANGELOG.md:53]: This heading is meant to also be a place for a point on https://github.com/jupyterlab/jupyterlab/pull/15927 (once it is merged)

Comment (krassowski) [reply CHANGELOG.md:54]: I believe non-native speaker (most of the world) will struggle to understand `JupyterLab-ify`. Also, these shortcuts work the same in File editor and other editors (settings editor, metadata form, etc) so saying "cell editor" is non acurate.

File: CHANGELOG.md [hunk 9]
```diff
@@ -4,6 +4,60 @@
 
 # JupyterLab Changelog
 
+## v4.2
+
+### Workspaces UI
+
+It is now possible to switch and manage [workspaces](https://jupyterlab.readthedocs.io/en/latest/user/workspaces.html) from within JupyterLab.
+
+<img alt="A context menu opened over the sidebar with workspaces list. The menu has options to manager the workspace over which it was opened - clone, rename, reset delete, export and import." src="https://raw.githubusercontent.com/krassowski/jupyterlab/ufc4.2/docs/source/getting_started/changelog_assets/4.2-workspaces-sidebar.png" class="jp-screenshot">
```
Comment (JasonWeill) [inline CHANGELOG.md:13]:
```suggestion
<img alt="A context menu opened over the sidebar with workspaces list. The menu has options to managerthe workspace over which it was opened: clone, rename, reset, delete, export, and import." src="https://raw.githubusercontent.com/krassowski/jupyterlab/ufc4.2/docs/source/getting_started/changelog_assets/4.2-workspaces-sidebar.png" class="jp-screenshot">
```

Copy edit on alt text — thank you for adding this!

File: CHANGELOG.md [hunk 10]
```diff
@@ -4,6 +4,60 @@
...
+
+- quickly jump to open tabs/recently closed files by using a new searchable modal dialog (press <kbd>Ctrl</kbd>+<kbd>Alt</kbd>+<kbd>A</kbd> to open the dialog, or click on the \[↗\] icon in the sidebar):
+
+   <img alt="A modal dialog for quickly jumping to open/closed files" src="https://raw.githubusercontent.com/krassowski/jupyterlab/ufc4.2/docs/source/getting_started/changelog_assets/4.2-recent-modal.png" class="jp-screenshot">
+
+### Full notebook windowing mode by default
+
+Notebooks in the `full` windowing mode only render the visible cells, significantly improving the performance of the application. The limitation of `full` mode is that the search function in your browser may produce false negatives - using the [JupyterLab search function](https://jupyterlab.readthedocs.io/en/latest/user/interface.html#searching) is recommended. To revert to the behaviour from JupyterLab 4.1 go to Settings → Settings Editor → Notebook, scroll to "Windowing mode" and choose `defer`.
```
Comment (JasonWeill) [inline CHANGELOG.md:35]:
```suggestion
Notebooks in the `full` windowing mode only render the visible cells, significantly improving the performance of the application. One limitation of `full` mode is that the search function in your browser may produce false negatives; using the [JupyterLab search function](https://jupyterlab.readthedocs.io/en/latest/user/interface.html#searching) is recommended. To revert to the behaviour from JupyterLab 4.1, go to Settings → Settings Editor → Notebook, scroll to "Windowing mode", and choose `defer`.
```

File: CHANGELOG.md [hunk 11]
```diff
@@ -4,6 +4,60 @@
...
+
+### Improved Shortcuts Editor
+
+Among numerous improvements and bug fixes for keyboard shortcuts editor:
```
Comment (JasonWeill) [inline CHANGELOG.md:39]:
```suggestion
Among the numerous improvements and bug fixes for the keyboard shortcuts editor:
```

File: CHANGELOG.md [hunk 12]
```diff
@@ -4,6 +4,60 @@
...
+- it is now possible to remove the default shortcuts,
+- shortcuts are correctly sorted when using a language pack,
+- shortcuts with different arguments are now correctly displayed as individual entries.
+
+<img alt="A context menu opened over a keybinding in the Keyboard Shortcuts setting panel within the Settings widget" src="https://raw.githubusercontent.com/krassowski/jupyterlab/ufc4.2/docs/source/getting_started/changelog_assets/4.2-delete-shortcut.png" class="jp-screenshot">
+
+### Dark high contrast theme
+
+A new theme, JupyterLab Dark High Contrast, which is intended to benefit users with the need for higher contrast, following the WCAG AAA accessibility standard for color contrast.
+To select this theme choose from the menu bar: Settings → Theme → JupyterLab Dark High Contrast. Please provide feedback and suggestions on further improvements to this theme.
```
Comment (JasonWeill) [inline CHANGELOG.md:50]:
```suggestion
To select this theme, from the menu bar, choose Settings → Theme → JupyterLab Dark High Contrast. Please provide feedback and suggestions on further improvements to this theme.
```

Should we suggest that people provide issues? Open Discourse threads? Chat in the Matrix/Element/Zulip/etc room?

**Review by JasonWeill — COMMENTED** [review]
Looks good! Thanks for adding this.

File: CHANGELOG.md (hunk 9 above)
Comment (krassowski) [reply CHANGELOG.md:13]: `managerthe` guess you meant to remove "r", not the space - I will edit you comment

Comment (krassowski) [reply CHANGELOG.md:13]: Haha, nope there can be no `:` - this is what I had initially but `myst` errors out if there is a colon in `alt`, see my earlier commit https://github.com/jupyterlab/jupyterlab/pull/16103/commits/0fcad9dc7f6f0e4cfb85b9954fe55ab7bc788899. I will edit to remove the colon again.

**Comment by krassowski** [comment]
> Should we suggest that people provide issues? Open Discourse threads? Chat in the Matrix/Element/Zulip/etc room?

For me the preference is for GitHub issues, as no one actively monitors other places. I mean I do but I prefer GitHub issues. If someone else is willing to monitor other places we can add as many as we want.

File: CHANGELOG.md [hunk 13]
```diff
@@ -4,6 +4,60 @@
...
+<img alt="A context menu opened over a keybinding in the Keyboard Shortcuts setting panel within the Settings widget" src="https://raw.githubusercontent.com/krassowski/jupyterlab/ufc4.2/docs/source/getting_started/changelog_assets/4.2-delete-shortcut.png" class="jp-screenshot">
+
+### Dark high contrast theme
+
+A new theme, JupyterLab Dark High Contrast, which is intended to benefit users with the need for higher contrast, following the WCAG AAA accessibility standard for color contrast.
+To select this theme, from the menu bar, choose Settings → Theme → JupyterLab Dark High Contrast. Please provide feedback and suggestions on further improvements to this theme.
+
+<img alt="A dark high contrast theme" src="https://raw.githubusercontent.com/krassowski/jupyterlab/ufc4.2/docs/source/getting_started/changelog_assets/4.2-dark-high-contrast.png" class="jp-screenshot">
+
+### Keyboard shortcuts improvements
```
Comment (gabalafou) [inline CHANGELOG.md:54]: Are you saying you want to keep it generic? I can come up with other ideas...

```suggestion
### Hoist and harmonize CodeMirror keyboard shortcuts
```

Comment (krassowski) [reply CHANGELOG.md:54]: But this still does not cover https://github.com/jupyterlab/jupyterlab/pull/1592. Can we live it as-is for the time being?

Comment (krassowski) [reply CHANGELOG.md:54]: So I do not want users who were burned by 4.1 and are skimming 4.2 release notes to disregard this section. I mean I still hope we can have https://github.com/jupyterlab/jupyterlab/pull/15927 as a priority.

Comment (gabalafou) [reply CHANGELOG.md:54]: That makes sense

**Review by gabalafou — APPROVED** [review]

**Pull request merged**